File: SRM144/BinaryCode.cs
// BEGIN CUT HERE

// END CUT HERE
using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;

public class BinaryCode {
    public string[] decode(string message) {
        string[] res = new string[2];
        int[] decodeInt = new int[message.Length+1];
        int[] encodeInt = new int[message.Length + 1];
        string str = null;

        decodeInt[0] = 0;
        for( int i = 0; i<message.Length;i++)
        {
            decodeInt[i + 1] = message[i] - '0';
        }

        encodeInt[0] = 0;
        encodeInt[1] = 0;
        for(int i = 1; i < message.Length; i++)
        {
            encodeInt[i + 1] = decodeInt[i] - encodeInt[i - 1] - encodeInt[i];
            if (encodeInt[i + 1] != 0 && encodeInt[i + 1] != 1)
            {
                res[0] = "NONE";
                break;
            }
        }

        if(res[0] != "NONE")
        {
            if (decodeInt[message.Length] != encodeInt[message.Length - 1] + encodeInt[message.Length]) 
                res[0] = "NONE";
            else
            {
                for(int i = 1; i<message.Length + 1;i++)
                    str += encodeInt[i].ToString();
                //Console.WriteLine(str);
                res[0] = str;
                str = null;
            }
        }

        encodeInt = new int[message.Length + 1];
        encodeInt[0] = 0;
        encodeInt[1] = 1;
        for (int i = 1; i < message.Length; i++)
        {
            encodeInt[i + 1] = decodeInt[i] - encodeInt[i - 1] - encodeInt[i];
            if (encodeInt[i + 1] != 0 && encodeInt[i + 1] != 1)
            {
                res[1] = "NONE";
                break;
            }
        }

        if (res[1] != "NONE")
        {
            if (decodeInt[message.Length] != encodeInt[message.Length - 1] + encodeInt[message.Length])
                res[1] = "NONE";
            else
            {
                for (int i = 1; i < message.Length + 1; i++)
                    str += encodeInt[i].ToString();
                res[1] = str;
                str = null;
            }
        }


        return res;
    }

// BEGIN CUT HERE
    public static void Main(String[] args) {
        try  {
            eq(0,(new BinaryCode()).decode("123210122"),new string[] { "011100011",  "NONE" });
            eq(1,(new BinaryCode()).decode("11"),new string[] { "01",  "10" });
            eq(2,(new BinaryCode()).decode("22111"),new string[] { "NONE",  "11001" });
            eq(3,(new BinaryCode()).decode("123210120"),new string[] { "NONE",  "NONE" });
            eq(4,(new BinaryCode()).decode("3"),new string[] { "NONE",  "NONE" });
            eq(5,(new BinaryCode()).decode("12221112222221112221111111112221111"),new string[] { "01101001101101001101001001001101001",  "10110010110110010110010010010110010" });
        } 
        catch( Exception exx)  {
            System.Console.WriteLine(exx);
            System.Console.WriteLine( exx.StackTrace);
        }
        System.Console.ReadKey();
    }
    private static void eq( int n, object have, object need) {
        if( eq( have, need ) ) {
            Console.WriteLine( "Case "+n+" passed." );
        } else {
            Console.Write( "Case "+n+" failed: expected " );
            print( need );
            Console.Write( ", received " );
            print( have );
            Console.WriteLine();
        }
    }
    private static void eq( int n, Array have, Array need) {
        if( have == null || have.Length != need.Length ) {
            Console.WriteLine("Case "+n+" failed: returned "+have.Length+" elements; expected "+need.Length+" elements.");
            print( have );
            print( need );
            return;
        }
        for( int i= 0; i < have.Length; i++ ) {
            if( ! eq( have.GetValue(i), need.GetValue(i) ) ) {
                Console.WriteLine( "Case "+n+" failed. Expected and returned array differ in position "+i );
                print( have );
                print( need );
                return;
            }
        }
        Console.WriteLine("Case "+n+" passed.");
    }
    private static bool eq( object a, object b ) {
        if ( a is double && b is double ) {
            return Math.Abs((double)a-(double)b) < 1E-9;
        } else {
            return a!=null && b!=null && a.Equals(b);
        }
    }
    private static void print( object a ) {
        if ( a is string ) {
            Console.Write("\"{0}\"", a);
        } else if ( a is long ) {
            Console.Write("{0}L", a);
        } else {
            Console.Write(a);
        }
    }
    private static void print( Array a ) {
        if ( a == null) {
            Console.WriteLine("<NULL>");
        }
        Console.Write('{');
        for ( int i= 0; i < a.Length; i++ ) {
            print( a.GetValue(i) );
            if( i != a.Length-1 ) {
                Console.Write(", ");
            }
        }
        Console.WriteLine( '}' );
    }
// END CUT HERE
}

File: SRM144/BinaryCode.cpp
// BEGIN CUT HERE

// END CUT HERE
#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <queue>
#include <set>
#include <map>
#include <cstdio>
#include <cstdlib>
#include <cctype>
#include <cmath>
using namespace std;

class BinaryCode {
public:
    vector <string> decode(string message) {
        vector <string> res;
		vector<int> decodeInt(message.length()+1);
		vector<int> encodeInt(message.length()+1);
		stringstream ss;

		decodeInt[0] = 0;
		for (int i = 0; i < (int)message.length(); i++)
		{
			decodeInt[i+1] = message.at(i) - '0';
		}

		encodeInt[0] = 0;
		encodeInt[1] = 0;
		for (int i = 1; i < message.length() ; i++)
		{
			encodeInt[i + 1] = decodeInt[i] - encodeInt[i - 1] - encodeInt[i];
			if (encodeInt[i + 1] != 0 && encodeInt[i + 1] != 1)
			{
				res.push_back("NONE");
				break;
			}
		}


		if (res.size() == 0)
		{
			if (decodeInt[message.length()] != encodeInt[message.length() - 1] + encodeInt[message.length()])
			{
				res.push_back("NONE");
			}
			else
			{
				for (int i = 1; i < message.length() + 1; i++)
					ss << encodeInt[i];

				res.push_back(ss.str());
			}
		}


		ss.str("");
		ss.clear();
		encodeInt.clear();
		encodeInt.resize(message.length() + 1);
		encodeInt[0] = 0;
		encodeInt[1] = 1;
		for (int i = 1; i < message.length() ; i++)
		{
			encodeInt[i + 1] = decodeInt[i] - encodeInt[i - 1] - encodeInt[i];
			if (encodeInt[i + 1] != 0 && encodeInt[i + 1] != 1)
			{
				res.push_back("NONE");
				break;
			}
		}


		if (res.size() == 1)
		{
			if (decodeInt[message.length()] != encodeInt[message.length() - 1] + encodeInt[message.length()])
			{
				res.push_back("NONE");
			}
			else
			{
				for (int i = 1; i < message.length() + 1; i++)
					ss << encodeInt[i];

				res.push_back(ss.str());
			}
		}
		return res;
    }

};

vector<string> split( const string& s, const string& delim =" " ) {
    vector<string> res;
    string t;
    for ( int i = 0 ; i != s.size() ; i++ ) {
        if ( delim.find( s[i] ) != string::npos ) {
            if ( !t.empty() ) {
                res.push_back( t );
                t = "";
            }
        } else {
            t += s[i];
        }
    }
    if ( !t.empty() ) {
        res.push_back(t);
    }
    return res;
}

vector<int> splitInt( const string& s, const string& delim =" " ) {
    vector<string> tok = split( s, delim );
    vector<int> res;
    for ( int i = 0 ; i != tok.size(); i++ )
        res.push_back( atoi( tok[i].c_str() ) );
    return res;
}

// BEGIN CUT HERE
#define ARRSIZE(x) (sizeof(x)/sizeof(x[0]))

template<typename T> void print( T a ) {
    cerr << a;
}
static void print( long long a ) {
    cerr << a << "L";
}
static void print( string a ) {
    cerr << '"' << a << '"';
}
template<typename T> void print( vector<T> a ) {
    cerr << "{";
    for ( int i = 0 ; i != a.size() ; i++ ) {
        if ( i != 0 ) cerr << ", ";
        print( a[i] );
    }
    cerr << "}" << endl;
}
template<typename T> void eq( int n, T have, T need ) {
    if ( have == need ) {
        cerr << "Case " << n << " passed." << endl;
    } else {
        cerr << "Case " << n << " failed: expected ";
        print( need );
        cerr << " received ";
        print( have );
        cerr << "." << endl;
    }
}
template<typename T> void eq( int n, vector<T> have, vector<T> need ) {
    if( have.size() != need.size() ) {
        cerr << "Case " << n << " failed: returned " << have.size() << " elements; expected " << need.size() << " elements.";
        print( have );
        print( need );
        return;
    }
    for( int i= 0; i < have.size(); i++ ) {
        if( have[i] != need[i] ) {
            cerr << "Case " << n << " failed. Expected and returned array differ in position " << i << ".";
            print( have );
            print( need );
            return;
        }
    }
    cerr << "Case " << n << " passed." << endl;
}
static void eq( int n, string have, string need ) {
    if ( have == need ) {
        cerr << "Case " << n << " passed." << endl;
    } else {
        cerr << "Case " << n << " failed: expected ";
        print( need );
        cerr << " received ";
        print( have );
        cerr << "." << endl;
    }
}
// END CUT HERE

// BEGIN CUT HERE
void main( int argc, char* argv[] ) {
    {
        string retrunValueARRAY[] = { "011100011",  "NONE" };
        vector <string> retrunValue( retrunValueARRAY, retrunValueARRAY+ARRSIZE(retrunValueARRAY) );
        BinaryCode theObject;
        eq(0, theObject.decode("123210122"),retrunValue);
    }
    {
        string retrunValueARRAY[] = { "01",  "10" };
        vector <string> retrunValue( retrunValueARRAY, retrunValueARRAY+ARRSIZE(retrunValueARRAY) );
        BinaryCode theObject;
        eq(1, theObject.decode("11"),retrunValue);
    }
    {
        string retrunValueARRAY[] = { "NONE",  "11001" };
        vector <string> retrunValue( retrunValueARRAY, retrunValueARRAY+ARRSIZE(retrunValueARRAY) );
        BinaryCode theObject;
        eq(2, theObject.decode("22111"),retrunValue);
    }
    {
        string retrunValueARRAY[] = { "NONE",  "NONE" };
        vector <string> retrunValue( retrunValueARRAY, retrunValueARRAY+ARRSIZE(retrunValueARRAY) );
        BinaryCode theObject;
        eq(3, theObject.decode("123210120"),retrunValue);
    }
    {
        string retrunValueARRAY[] = { "NONE",  "NONE" };
        vector <string> retrunValue( retrunValueARRAY, retrunValueARRAY+ARRSIZE(retrunValueARRAY) );
        BinaryCode theObject;
        eq(4, theObject.decode("3"),retrunValue);
    }
    {
        string retrunValueARRAY[] = { "01101001101101001101001001001101001",  "10110010110110010110010010010110010" };
        vector <string> retrunValue( retrunValueARRAY, retrunValueARRAY+ARRSIZE(retrunValueARRAY) );
        BinaryCode theObject;
        eq(5, theObject.decode("12221112222221112221111111112221111"),retrunValue);
    }
}
// END CUT HERE
